File: frontend/src/types/workout.ts
export interface Exercise {
  name: string;
  sets?: number;
  reps?: number;
  weight?: number;
  duration?: number; // in seconds
}

export interface WorkoutTemplate {
  id: string;
  name: string;
  description?: string;
  mode: 'gym' | 'home';
  exercises: Exercise[];
  is_system: boolean;
  created_by?: string;
  created_at: string;
  updated_at: string;
}

export interface WorkoutSession {
  id: string;
  template_id?: string;
  name: string;
  mode: 'gym' | 'home';
  exercises: Exercise[];
  start_time: string;
  end_time?: string;
  duration?: number; // in seconds
  total_volume?: number;
  notes?: string;
  created_at: string;
}

export interface WorkoutStats {
  weekly: {
    sessions_count: number;
    total_duration: number; // in seconds
    total_volume: number;
    week_start: string;
  };
}

export interface CreateWorkoutTemplateRequest {
  name: string;
  description?: string;
  mode: 'gym' | 'home';
  exercises: Exercise[];
}

export interface CreateWorkoutSessionRequest {
  template_id?: string;
  name: string;
  mode: 'gym' | 'home';
  exercises: Exercise[];
  start_time: string;
  end_time?: string;
  notes?: string;
}

File: backend/app/models/workout.py
from sqlalchemy import Column, String, DateTime, ForeignKey, Enum, Text, JSON, Integer
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
import uuid
import enum
from ..core.database import Base

class WorkoutType(str, enum.Enum):
    gym = "gym"
    home = "home"

class WorkoutTemplate(Base):
    __tablename__ = "workout_templates"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    owner_user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)  # Null for system templates
    name = Column(String, nullable=False)
    type = Column(Enum(WorkoutType), nullable=False)
    exercises = Column(JSON, nullable=False)  # [{"name": str, "sets": int, "reps": int, "weight_kg": float?, "duration_sec": int?}]
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    # Relationships
    owner = relationship("User", back_populates="workout_templates")
    sessions = relationship("WorkoutSession", back_populates="template")

class WorkoutSession(Base):
    __tablename__ = "workout_sessions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    couple_id = Column(UUID(as_uuid=True), ForeignKey("couples.id"), nullable=True)
    template_id = Column(UUID(as_uuid=True), ForeignKey("workout_templates.id"), nullable=True)
    mode = Column(Enum(WorkoutType), nullable=False)
    start_time = Column(DateTime(timezone=True), nullable=False)
    end_time = Column(DateTime(timezone=True), nullable=True)
    notes = Column(Text, nullable=True)
    metrics = Column(JSON, nullable=True)  # {"total_volume": float, "calories_est": int?, "heart_rate_avg": int?}
    exercises_performed = Column(JSON, nullable=True)  # Actual exercises with completed sets/reps
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    # Relationships
    user = relationship("User", back_populates="workout_sessions")
    couple = relationship("Couple", back_populates="workout_sessions")
    template = relationship("WorkoutTemplate", back_populates="sessions")